Tomer/compressionable data enable

tomerr-lb opened the pull request.

yael-lb committed:
memtier: add crc_verify argument to memtier

add crc_verify argument:
this boolean argument enables data verification using a crc
function.
In this commit:
- add crc_verify parameter to memtier config
- add parsing of crc_verify argument
- add printing of crc_verify value in:
	- console configuration print function
	- json configuration print function

Issue: LBM1-3307

yael-lb committed:
memtier: crc_verify SET:GET ratio

In the current implementation of the crc verify, the only SET:GET
ratio that is supported is 1:0. Thus, ratio configuration is
disallowed and the ratio is always set to 1:0.
This determines the request pattern when running the benchmark -
only SET requests will be sent.
In the data verification phase, a GET request would be generated
for each of the keys that were set and the returned value
will be verified.

Issue: LBM1-3307

yael-lb committed:
memtier: add crc32 static class

Issue: LBM1-3307

yael-lb committed:
memtier: new crc object generating class

Add class crc_object_generator that inherits from class
object_generator.
Main functionality differnces from ancestor class:
- Generates objects with values that have a crc that was calculated
  on them and resides at their tail.

Issue: LBM1-3307

yael-lb committed:
memtier: use crc object generator

- Use crc object generator when crc-verify argument was passed to
  memtier.
- Verify benchmark parameters.

Issue: LBM1-3307

yael-lb committed:
memtier: new crc verfiy client class

Add class crc_verify_client that inherits from class client.
Main functionality differnces from ancestor Class:
- Generate GET request against every SET request that was performed
  when running the benchmark.
- Perform crc verification on the received responses values.

Issue: LBM1-3307

yael-lb committed:
memtier: verify data

Run data verification check when crc-verify argument was
passed to memtier.

Issue: LBM1-3307

adir-lb commented on protocol.cpp:
```diff
@@ -190,12 +192,11 @@ int redis_protocol::authenticate(const char *credentials)
     assert(key_len > 0);
-    assert(value != NULL);
-    assert(value_len > 0);
+    assert(total_buffers_len > 0);
```
don't you need to assert each buffer within the list?

adir-lb commented on protocol.cpp:
```diff
@@ -240,11 +241,16 @@ int redis_protocol::write_command_set(const char *key, int key_len, const char *
+    }
+
+    for (val_list_it it = values_list->begin(); it != values_list->end(); ++it)
+    {
```
what is the convention for one-liner for loop?
anyway i think they use to put the `{` in the same line of the `for` statement.

sashas-lb commented on obj_gen.cpp:
```diff
@@ -454,9 +455,13 @@ void data_object::clear(void)
+    m_total_buffers_len = 0;
     m_expiry = 0;
+    if (m_values_list)
+    {
```
``` c 
 if (m_values_list)  {
```

adir-lb commented on obj_gen.h:
```diff
@@ -64,8 +67,9 @@ class data_object {
     const char* get_key(unsigned int* key_len);
-    void set_value(const char* value, unsigned int value_len);
-    const char* get_value(unsigned int* value_len);
+    void add_value(const char* value, unsigned int value_leni, bool clear_list = false);
```
`value_leni`?

adir-lb replied:
`clear_list` doesn't seem related to this function.

adir-lb commented on protocol.h:
```diff
@@ -20,6 +20,10 @@
+#include <utility>
+#include <list>
+
+typedef std::list<std::pair<const char*, unsigned int> > val_list;
```
maybe your new type should contain also the total value size.

adir-lb commented on memtier_benchmark.cpp:
```diff
@@ -90,6 +90,7 @@ static void config_print(FILE *file, struct benchmark_config *cfg)
         "clients = %u\n"
         "threads = %u\n"
         "cpu_split = %s\n"
+        "compress_perc = %d\n"
```
consider calling this param compression_ratio.

adir-lb replied:
btw, should it be `%u`?

tomerr-lb replied:
yes it should

adir-lb commented on obj_gen.cpp:
```diff
@@ -221,7 +223,14 @@ void object_generator::alloc_value_buffer(void)
+        float decimal = static_cast<float>(m_compress_perc)/100;
+        m_value_buffer_size = static_cast<unsigned int>(size*(1-decimal));
+        m_zeros_buffer_size = size - m_value_buffer_size;
+        m_zeros_buffer = (char*) calloc(m_zeros_buffer_size,1);
```
cosmetic: please add spaces around operators (`/`, `-`. `*`) and `,`.

adir-lb commented on obj_gen.cpp:
```diff
@@ -221,7 +223,14 @@ void object_generator::alloc_value_buffer(void)
+    if (!m_compress_perc)
+        m_value_buffer_size = size;
+    else {
+        float decimal = static_cast<float>(m_compress_perc)/100;
```
consider the following - 
```cpp
float decimal = m_compress_perc / 100.0;
```

tomerr-lb committed:
memtier: cosmetics

tomerr-lb committed:
memtier: Protocols now expects list of buffer and size pairs

This makes it possible to pipe several buffers into one set command,
 which enables manipulating the data sent by memtier

Issue: LBM1-3452

yael-lb committed:
memtier: add verified_key and errors to run_stats totals

Issue: LBM1-3307

yael-lb committed:
memtier: new class verify_client_group

Add class verify_client_group that inherits from class client_group.
Main functionality differnces from ancestor Class:
- Create client group with crc_clients
- Merge total verification stats: verified keys and errors.

Issue: LBM1-3307

yael-lb committed:
memtier: enable running benchmarks with crc_verify clients

Issue: LBM1-3307

tomerr-lb committed:
memtier: new crc verfiy client class

Add class crc_verify_client that inherits from class client.
Main functionality differnces from ancestor Class:
- Generate GET request against every SET request that was performed
  when running the benchmark.
- Perform crc verification on the received responses values.

Issue: LBM1-3307

tomerr-lb committed:
memtier: enabled to set data compressibility

Added compress-perc parameter to the data compressibility percentiles,
setting this parameter will set the memtier generated data to be at least compress-perc compressible

Issue: LBM1-3452
